crates/ra_hir/src/ty.rs

```diff
@@ -89,13 +76,30 @@ pub enum Ty {
     /// fn foo() -> i32 { 1 }
     /// let bar: fn() -> i32 = foo;
     /// ```
-    FnPtr(FnSig),
+    FnPtr,
 
     /// The never type `!`.
     Never,
 
     /// A tuple type.  For example, `(i32, bool)`.
-    Tuple(Arc<[Ty]>),
+    Tuple,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct ApplicationTy {
+    pub name: TypeName,
+    pub parameters: Substs,
+}
+
+/// A type. This is based on the `TyKind` enum in rustc (librustc/ty/sty.rs).
+///
+/// This should be cheap to clone.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Ty {
+    /// A nominal type with (maybe 0) type parameters. This might be a primitive
+    /// type like `bool`, a struct, tuple, function pointer, reference or
+    /// several other things.
+    Apply(ApplicationTy),
```

Comment:
Do we need a separate `ApplicationTy` type? If we remove it, matching will become simpler. If we keep it, than perhaps we could use the same name for enum variant and strut? (`Apply(ApplyTy)`)

Reply:
I don't think *we* need it, but Chalk has it... so the question is here as well why it is that way in Chalk :)

Reply:
With as_adt and the like, the point about nested matching becomes moot